File: energielenker/energielenker/work_order/work_order.js
// Copyright (c) 2024, libracore and contributors
// For license information, please see license.txt

frappe.ui.form.on('Work Order', {
    refresh: function(frm) {
        if (frm.doc.__islocal) {
            if (frm.doc.bom_no) {
                set_bom_values(frm);
            }
        } else {
            check_not_transferred_items(frm);
        }
        
        if (frm.doc.status == "In Process") {
            
        }
    }
});

function set_bom_values(frm) {
    frappe.call({
        'method': "energielenker.energielenker.work_order.work_order.get_bom_values",
        'args': {
            'bom': frm.doc.bom_no
        },
        'callback': function(response) {
            var item = response.message[0].item;
            var sales_order = response.message[0].sales_order;
            var project = response.message[0].project;
            var fg_warehouse = response.message[0].default_warehouse_readonly;
            var wip_warehouse = response.message[0].depot_warehouse;
            var item_so_detail = response.message[0].item_so_detail;
            cur_frm.set_value("production_item", item);
            setTimeout(function(){
                cur_frm.set_value("project", project);
                cur_frm.set_value("fg_warehouse", fg_warehouse);
                cur_frm.set_value("wip_warehouse", wip_warehouse);
                cur_frm.set_value("use_multi_level_bom", 0);
                cur_frm.set_value("item_so_detail", item_so_detail);
            }, 1000);
            setTimeout(function(){
                cur_frm.set_value("sales_order", sales_order);
            }, 2000);
        }
    });
}

function check_not_transferred_items(frm) {
    if (frm.doc.status == "In Process" || frm.doc.status == "Stopped" || frm.doc.status == "Completed") {
        let affected_items = []
        for (let i = 0; i < frm.doc.required_items.length; i++) {
            if (frm.doc.required_items[i].required_qty > frm.doc.required_items[i].transferred_qty) {
                affected_items.push(frm.doc.required_items[i].item_code);
            }
        }
        if (affected_items.length > 0) {
            cur_frm.dashboard.add_comment("<b>ACHTUNG! Es ist noch nicht alles kommissioniert. Bitte VOR der Fertigmeldung alles kommissionieren!<br>Artikel " + affected_items.join(', ') + " sind noch nicht komissioniert.</b>", 'green', true);
            if (frm.doc.status == "In Process") {
                frm.page.remove_inner_button('Finish');
            }
        }
    }
}

File: energielenker/energielenker/work_order/work_order.py
# -*- coding: utf-8 -*-
# Copyright (c) 2024, libracore AG and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe

@frappe.whitelist()
def get_bom_values(bom):
    depot = frappe.db.get_value("energielenker Settings", "energielenker Settings", "depot_warehouse")
    data = frappe.db.sql("""SELECT
                    `tabBOM`.`item`,
                    `tabBOM`.`sales_order`,
                    `tabBOM`.`project`,
                    `tabBOM`.`item_so_detail`,
                    `tabItem`.`default_warehouse_readonly`,
                    "{depot}" AS depot_warehouse
                FROM `tabBOM`
                LEFT JOIN `tabItem` ON `tabItem`.`name` = `tabBOM`.`item`
                WHERE `tabBOM`.`name` = '{bom}'""".format(bom=bom, depot=depot), as_dict=True)
                
    
    return data

@frappe.whitelist()
def set_has_work_order_in_sales_order(doc, method):
    try:
        frappe.db.sql("""UPDATE `tabSales Order` SET `has_work_order` = 1 WHERE `name` = '{sales_order}'""".format(sales_order=doc.sales_order))
    except Exception as e:
        frappe.log_error(str(e), "set_has_work_order_in_sales_order")
    return

@frappe.whitelist()
def unset_has_work_order_in_sales_order(doc, method):
    try:
        #first check if there are any other uncancelled work orders for this sales order
        work_orders = frappe.get_all("Work Order", filters={"sales_order": doc.sales_order, "docstatus": 1}, fields=["name"])
        if len(work_orders) > 0:
            return
        
        frappe.db.sql("""UPDATE `tabSales Order` SET `has_work_order` = 0 WHERE `name` = '{sales_order}'""".format(sales_order=doc.sales_order))
    except Exception as e:
        frappe.log_error(str(e), "unset_has_work_order_in_sales_order")
    return